### PR TITLE
Standard return value codes. 

### PR DESCRIPTION
**Pull request by phil-levis**
These allow operations to signify whether they were performed successfully or a run-time error occurred. As we move towards standardizing APIs and error condition handling, these will be needed.

### CODE REVIEW

File: kernel/src/common/result.rs [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+//! success or an error code.
+//!
+//!  Author: Philip Levis <pal@cs.stanford.edu>
+//!  Date: Nov 21, 2016
+
+enum Result {
+    SUCCESS,
+    FAIL, // Generic failure condition
+    EBUSY, // Underlying system is busy; retry
+    EALREADY, // The state requested is already set
```
Comment (ppannuto) [inline kernel/src/common/result.rs]: Is there a good example of a situation where you wouldn't want to simply return SUCCESS instead of EALREADY?

Comment (alevy) [inline kernel/src/common/result.rs]: @ppannuto do you mean because normally we expect the operation to be idempotent so why return an error?

Comment (ppannuto) [inline kernel/src/common/result.rs]: Precisely

File: kernel/src/common/result.rs [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    EOFF, // The component is powered down
```
Comment (ppannuto) [inline kernel/src/common/result.rs]: Feeling bold, do we actually need this, or should a component just power-on on demand?

**Comment by alevy** [comment]
@ppannuto @phil-levis a reasonable strategy to me seems like either:

1. include as many things as we can think of reasonably, start using them, then maybe remove ones that end up being redundant.

2. include only the bare essentials and extend as needed

Either seems fine with me, personally

File: kernel/src/common/result.rs (hunk 1 above)
Comment (phil-levis) [inline kernel/src/common/result.rs]: The state transition signals an event to denote it has completed. E.g., turning something on has an event that denotes it's on. If you return SUCCESS, that implies an event will be signaled later. If the device is already on, then signaling the event requires having an interrupt source (invoking the callback from within the call isn't what the caller expects). There are therefore cases where the device needs to be a timer client in order to meet the interface.

That being said, I agree that this is an edge case very far from what we've dealing with right now, so I'm fine leaving it out.

**Comment by ppannuto** [comment]
I lean towards (2) -- narrow interfaces being easier to comprehend and maintain; removal of cruft happens rarely at best in practice --, but certainly don't feel strongly about two `#defines`

File: kernel/src/common/result.rs (hunk 2 above)
Comment (phil-levis) [inline kernel/src/common/result.rs]: Sometimes yes, but not always. E.g., if the RF233 is in DEEP_SLEEP, then trying to set the channel shouldn't turn the radio on.

**Comment by phil-levis** [comment]
I added this because I'd like to start using it in an RF233 stack implemented in Rust. I included return codes I thought very likely we'll need in the next 6-9 months or so, but am happy to scale it back to ones I'll need very soon. Those would be FAIL, EBUSY, EOFF, EALREADY, and ESIZE. If we have locks for the SPI, then there will need to be ERESERVE.

**Comment by ppannuto** [comment]
I'm pretty convinced by the use cases here. The two I commented on were the ones that jumped out at us as potentially wonky here when we read the list in lab.

I do think there's some author education that's worth putting out there, i.e. returning `SUCCESS` if a callback isn't expected, and handling power state automatically if it isn't an energy-expensive operation, but all of these here look are fine.

**Comment by alevy** [comment]
I say we go with these for now.

One thing to keep a lookout for is:
https://github.com/domrizzo/tock-hotel/pull/31/files#diff-be22de6e05181aecc93326a1be254b69 in domrizzo/tock-hotel#31

Perhaps it would be useful to start a thread on the mailing list to propose/coordinate?

**Pull request merged**